File: spkmeans.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "spkmeans.h"
#include "spkmeanslogic.c"
#include "kmeans.c"



#define EPSILON 0.001
#define assertAndReturn(cond)                   \
        if (!(cond)) {                          \
            printf("An Error Has Occurred");    \
            exit(1);                           \
        }                                       \


/*
struct cord
{
    double value;
    struct cord *next;
};
struct vector
{
    struct vector *next;
    struct cord *cords;
};


double** convertLinkedListToArray(struct vector *v, int vectors, int cords);
void freeLinkedList(struct vector *vec);
void freeCordsList(struct cord *cord);
*/
char *goal, *file;
int rowsCount=0;
int columnCount=0;
double **data_points =NULL;

int main(int argc, char *argv[] )
{
    if(argc != 2){  // 3
        printf("An Error Has Occurred");
        return 0;
    }
    goal = argv[1];
    //file = argv[2];
    file = "./inputj.txt";
    prepData(file);

    printf("%d %d",rowsCount,columnCount);

    if(strcmp(goal,"wam")==0){
        double **wamRes = wam(data_points, rowsCount,columnCount);
        printMatrix(wamRes, rowsCount, rowsCount);
        freeMatrix(wamRes,rowsCount);
    }
    else if(strcmp(goal,"ddg")==0){ 
        double **wamRes = wam(data_points, rowsCount,columnCount);
        double **ddgRes = ddg(wamRes,rowsCount);
        printMatrix(ddgRes, rowsCount, rowsCount);
        freeMatrix(wamRes,rowsCount);
        freeMatrix(ddgRes,rowsCount);
    }
    else if(strcmp(goal,"gl")==0){
        double **wamRes = wam(data_points, rowsCount,columnCount);
        double **ddgRes = ddg(wamRes,rowsCount);
        double **glRes = gl(wamRes,ddgRes,rowsCount);
        printMatrix(glRes, rowsCount, rowsCount);
        freeMatrix(wamRes,rowsCount);
        freeMatrix(ddgRes,rowsCount);
        freeMatrix(glRes,rowsCount);
    } 
    else if(strcmp(goal,"jacobi")==0){
        double ***jacobiRes;
        assertAndReturn(rowsCount==columnCount);
        jacobiRes = jacobi(data_points, rowsCount);
        printDiag(jacobiRes[0], rowsCount);
        printf("\n");
        printMatrix(jacobiRes[1], rowsCount, rowsCount);

        freeMatrix(jacobiRes[1],rowsCount);
        freeMatrix(jacobiRes[0],rowsCount);
        free(jacobiRes);

    }
    else {
        printf("An Error Has Occurred");
        return 0;
    }
    return 1;
}


void prepData(char *filename){

    double n;
    char c;
    int columnsCountInitialized=0;

    struct vector *head_vec, *curr_vec,*prev_vec;
    struct cord *head_cord, *curr_cord;

    FILE *f = fopen(filename, "r");
    assertAndReturn(f!=NULL);
    
    head_cord = malloc(sizeof(struct cord));
    curr_cord = head_cord;
    curr_cord->next = NULL;

    head_vec = malloc(sizeof(struct vector));
    curr_vec = head_vec;
    curr_vec->next = NULL;


    while (fscanf(f,"%lf%c", &n, &c) == 2)
    {
        if(!columnsCountInitialized){
            columnCount++;
        }

        if (c == '\n' || c=='\r')
        {
            columnsCountInitialized=1;
            curr_cord->value = n;
            curr_vec->cords = head_cord;
            curr_vec->next = malloc(sizeof(struct vector));
            prev_vec=curr_vec;
            curr_vec = curr_vec->next;
            curr_vec->next = NULL;
            head_cord = malloc(sizeof(struct cord));
            curr_cord = head_cord;
            curr_cord->next = NULL;
            rowsCount++;
            continue;
        }

        curr_cord->value = n;
        curr_cord->next = malloc(sizeof(struct cord));
        curr_cord = curr_cord->next;
        curr_cord->next = NULL;

    }
    prev_vec->next = NULL;
    free(head_cord); 
    free(curr_vec); 
    fclose(f);
    data_points = convertLinkedListToArray(head_vec, rowsCount,columnCount);
    freeLinkedList(head_vec);

}


/*
double** convertLinkedListToArray(struct vector *v, int vectors, int cords){
    int i,j;
    double **res=malloc(vectors*sizeof(double*)) ;
    assertAndReturn(res!=NULL);

    for (i=1;i<=vectors;i++){
        struct cord *currCord= v->cords;
        res[i-1] = malloc(cords*sizeof(double));
        assertAndReturn(res[i-1]!=NULL);

        for (j=1;j<=cords;j++){
            res[i-1][j-1]=currCord->value;
            if(j<cords){
                currCord=currCord->next;
            }
        }
        if(i<vectors){
            v=v->next;
        }
    }
    return res;
}
void freeCordsList(struct cord *cord){
    if(cord != NULL){
        freeCordsList(cord->next);
        free(cord);
    }
}
void freeLinkedList(struct vector *vec){
    if(vec!= NULL){

        freeLinkedList(vec->next);

        freeCordsList (vec->cords); 

        free(vec);
    }
}

*/

File: spkmeans.py
import sys
import math
import os
import numpy as np
import pandas as pd
import random
import mykmeanssp as km
script_dir = os.path.dirname(__file__)


np.random.seed(0)

def Transpose_mat (A, n, m):
        AT = []
        inner_list = []
        for i in range(m):
            for j in range(n):
                inner_list.append(A[j][i])
                AT.append(inner_list)
        return AT

def readingUserArgs():
    if len(sys.argv)!=4 and len(sys.argv)!= 3:
        print("An Error Has Occurred")
        exit()
    if len(sys.argv) == 4:
        k, goal, file_name = sys.argv[1], sys.argv[2], sys.argv[3]
    else:
        goal, file_name = sys.argv[1], sys.argv[2]
        k = 1; #defult - need uristic  


    if file_name.endswith('txt'):
        data = np.genfromtxt(file_name,dtype=float,delimiter=",")
    else:
        with open(file_name, 'r', encoding='utf-8-sig') as f: 
            data = np.genfromtxt(f, dtype=float, delimiter=',')
    
    '''
    def prepData(data):
        res = []
        for line in data.readlines():
            innerlist = []
            for val in line.split(","):
                innerlist.append(float(val))
            res.append(innerlist)
        return res
    '''
    
    data_df=pd.DataFrame(data)

    rowCount = data_df.shape[0]
    columnCount = data_df.shape[1]
    
    
    if len(sys.argv) == 4:
        if not (k).isnumeric():
            print("Invalid number of clusters!")
            sys.exit()
        k = int(k)
        if (k) % 1 != 0 or k <= 1 or k >= rowCount:
            print("Invalid number of clusters!")
            sys.exit()
    

    x=np.array(data_df)
    data_p= [[x[i][j] for j in range(columnCount)]  for i in range(rowCount)]


    def kFromEigengapHeuristic (eigenvalues):
        max_delta = 0
        k = 1
        for i in range(len(eigenvalues)/2):
            delta_i = abs(eigenvalues[i] - eigenvalues[i+1])
            if delta_i > max_delta:
                max_delta = delta_i
                k = i
        return k
    
    
    
    
    def printMat(mat):
        for row in mat:
            print(','.join('{:.4f}'.format(v) for v in row))

    def printMatJacobi(mat,n):
        eigenvalues = mat[-1]
        print(','.join('{:.4f}'.format(val) for val in eigenvalues))
        for row in mat[:n]:  
            print(','.join('{:.4f}'.format(v) for v in row))

    
    if goal == "spk":
        matRes = km.jacobi(data_p,rowCount)
        eigenvalues = matRes[rowCount-1]

        if k==1:
            k = kFromEigengapHeuristic(eigenvalues)
        
        matResTrans = Transpose_mat(matRes, rowCount+1, columnCount)
        matResTrans_sorted = sorted(matRes , key=lambda k: k[-1])  # sort by the last element - eigenvalues
        U_Trans = matResTrans_sorted[:k]    # first k rows
        
        # remove last element in U in each row
        U = [row[:rowCount] for row in U]
        U = Transpose_mat(U_Trans, k, rowCount)     # U size = n * k

        # send U to kmeans++ 
        calcKmeans(U, rowCount, k, k)
    
    elif goal == "wam":
        wamRes = km.wam(data_p,rowCount,columnCount)  
        printMat(wamRes)
    elif goal == "ddg":
        wamRes = km.wam(data_p,rowCount,columnCount)
        ddgRes = km.ddg(wamRes,rowCount)
        printMat(ddgRes)
    elif goal == "gl":
        wamRes = km.wam(data_p,rowCount,columnCount)
        ddgRes = km.ddg(wamRes,rowCount)
        glRes = km.gl(wamRes, ddgRes,rowCount)
        printMat(glRes)

    elif goal == "jacobi":
        matRes = km.jacobi(data_p, rowCount)
        printMatJacobi(matRes, rowCount)

    else:
        print("An Error Has Occurred")
        sys.exit()


    
    




def calcKmeans(U, rowCount, columnCount, k):

    def d(dataItem,centroid):
        dataItem = dataItem[1:]
        centroid = centroid[1:]
        res = np.sqrt(np.sum((dataItem-centroid) ** 2))
        return res

    iter = 300     # not sure
    epsilon = 0    # not sure
    data_points = pd.DataFrame(U)

    data_points.columns = [str(i) for i in range(data_points.shape[1])]
    data_points = data_points.sort_values('0')
    
    randomRow = np.random.choice(data_points['0'], 1)[0]
    newCentroid = np.array(data_points[data_points['0']==randomRow])
    centroids_indices = [int(newCentroid[0][0])]
    centroids = np.copy(newCentroid)

    while (len(centroids) < k):

        prob = np.zeros(rowCount)
        for i in range(rowCount):
            x_i = data_points.iloc[i].to_numpy()
            minD = d(x_i , centroids[0])
            for centroid in centroids:
                currD = d(x_i , centroid)
                if currD < minD:
                    minD = currD

            prob[i] = minD

        if np.sum(prob) != 0:
            prob = prob / np.sum(prob)


        randomRow = np.random.choice(data_points['0'], 1 , p = prob)[0]
        newCentroid = np.array(data_points[data_points['0']==randomRow])

        centroids = np.append(centroids,newCentroid, axis=0)
        centroids_indices.append(int(newCentroid[0][0]))


    # print(centroids)
    x=np.array(data_points)
    data_p= [[x[i][j] for j in range(1,columnCount)]  for i in range(rowCount)]
    cent_p= [[centroids[i][j] for j in range(1,columnCount)]  for i in range(k)]

    calcCentroids = km.spk(cent_p,data_p,k,iter,epsilon,rowCount,columnCount)  
    print(','.join(str(v) for v in centroids_indices))
    for centroid in calcCentroids:
        print(','.join('{:.4f}'.format(v) for v in centroid))


readingUserArgs()    